### PR TITLE
Support 16bit activations

### PR DESCRIPTION
**Pull request by elad-c**
## Pull Request Description:
This PR add input bit-width information to the TPC (IMX500.v4). Each operator may support different input bit-widths. MCT will filter a node's qunatization config candidates according to the following nodes input bit-width capability.

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v4/tp_model.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    # 16 bits inputs and outputs
+    default_config_input16 = default_config.clone_and_edit(
+        supported_input_activation_n_bits=(8, 16))
+    default_config_input16_output16 = default_config_input16.clone_and_edit(
+        activation_n_bits=16, force_signedness=True)
+    default_configuration_options_inout16 = tp.QuantizationConfigOptions([default_config_input16_output16,
+                                                                          default_config_input16],
+                                                                         base_config=default_config_input16)
+    # 16 bits inputs and outputs
+    const_config_input16 = const_config.clone_and_edit(
```
Comment (haihabi) [inline model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v4/tp_model.py:179]: bineary_operation_with_16_bits

File: model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v4/tp_model.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    # If the QuantizationConfigOptions contains only one configuration,
+    # this configuration will be used for the operation quantization:
+    default_configuration_options = tp.QuantizationConfigOptions([default_config])
+
+    # Create a QuantizationConfigOptions for quantizing constants in functional ops.
+    # Constant configuration is similar to the default eight bit configuration except for PoT
+    # quantization method for the constant.
+    # Since the constants are not named attributes of the layer, we use the default_weight_attr_config to
+    # define the desired quantization properties for them.
+    const_config = default_config.clone_and_edit(
```
Comment (haihabi) [inline model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v4/tp_model.py:164]: default_with_const

File: model_compression_toolkit/core/common/quantization/set_node_quantization_config.py [hunk 3]
```diff
@@ -198,17 +205,46 @@ def _create_node_candidates_qc(qc: QuantizationConfig,
...
+        tpc: TargetPlatformCapabilities to get default OpQuantizationConfig.
         mixed_precision_enable: is mixed precision enabled
 
     Returns:
         List of candidates of weights quantization configurations to set for a node.
     """
 
+    # Filter quantization config options that don't match the graph.
+    _base_config = node_qc_options.base_config
+    _node_qc_options = node_qc_options.quantization_config_list
```
Comment (haihabi) [inline model_compression_toolkit/core/common/quantization/set_node_quantization_config.py:218]: Wrap in a function filter quantization_config_list

File: model_compression_toolkit/core/common/quantization/set_node_quantization_config.py [hunk 4]
```diff
@@ -198,17 +205,46 @@ def _create_node_candidates_qc(qc: QuantizationConfig,
...
         node: A node to set quantization configuration candidates to.
+        graph: Model's internal representation graph.
+        tpc: TargetPlatformCapabilities to get default OpQuantizationConfig.
         mixed_precision_enable: is mixed precision enabled
 
     Returns:
         List of candidates of weights quantization configurations to set for a node.
     """
 
+    # Filter quantization config options that don't match the graph.
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/quantization/set_node_quantization_config.py:216]: add comment with the criterion to eliminate config options

File: model_compression_toolkit/target_platform_capabilities/target_platform/op_quantization_config.py [hunk 5]
```diff
@@ -193,9 +204,21 @@ def __eq__(self, other):
...
             self.activation_n_bits == other.activation_n_bits and \
+            self.supported_input_activation_n_bits == other.supported_input_activation_n_bits and \
             self.enable_activation_quantization == other.enable_activation_quantization and \
+            self.force_signedness == other.force_signedness and \
             self.simd_size == other.simd_size
 
+    @property
+    def max_input_activation_n_bits(self) -> int:
+        """
+        Get maximum supported bit-width.
```
Comment (ofirgo) [inline model_compression_toolkit/target_platform_capabilities/target_platform/op_quantization_config.py:215]: of input (here and in the return + add '.')

File: model_compression_toolkit/core/common/quantization/set_node_quantization_config.py [hunk 6]
```diff
@@ -198,17 +205,46 @@ def _create_node_candidates_qc(qc: QuantizationConfig,
...
+        next_nodes_qc_options = [_node.get_qco(tpc) for _node in next_nodes]
+        next_nodes_supported_input_bitwidth = min([op_cfg.max_input_activation_n_bits
+                                                   for qc_opts in next_nodes_qc_options
+                                                   for op_cfg in qc_opts.quantization_config_list])
+
+        # Filter node's QC options that match next nodes input bit-width.
+        _node_qc_options = [_option for _option in _node_qc_options
+                            if _option.activation_n_bits <= next_nodes_supported_input_bitwidth]
+        if len(_node_qc_options) == 0:
+            Logger.critical(f"Graph doesn't match TPC bit configurations: {node} -> {next_nodes}")
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/quantization/set_node_quantization_config.py:230]: Add pragma: no cover if there is no test for this error (relevant for all exceptions)

File: model_compression_toolkit/core/common/quantization/set_node_quantization_config.py [hunk 7]
```diff
@@ -198,17 +205,46 @@ def _create_node_candidates_qc(qc: QuantizationConfig,
...
+
+        # Verify base config match
+        if any([node_qc_options.base_config.activation_n_bits > qc_opt.base_config.max_input_activation_n_bits
+                for qc_opt in next_nodes_qc_options]):
+            # base_config activation bits doesn't match next node supported input bit-width -> replace with
+            # a qco from quantization_config_list with maximum activation bit-width.
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/quantization/set_node_quantization_config.py:236]: add a warning with this (or similar) message

File: model_compression_toolkit/core/common/quantization/set_node_quantization_config.py [hunk 8]
```diff
@@ -198,17 +205,46 @@ def _create_node_candidates_qc(qc: QuantizationConfig,
...
+            if len(_node_qc_options) > 0:
+                output_act_bitwidth = {qco.activation_n_bits: i for i, qco in enumerate(_node_qc_options)}
+                _base_config = _node_qc_options[output_act_bitwidth[max(output_act_bitwidth)]]
+            else:
+                Logger.critical(f"Graph doesn't match TPC bit configurations: {node} -> {next_nodes}")
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/quantization/set_node_quantization_config.py:241]: add '.'

File: tests/common_tests/helpers/generate_test_tp_model.py [hunk 9]
```diff
@@ -30,6 +30,9 @@
 
 
 def generate_test_tp_model(edit_params_dict, name=""):
+    # Add "supported_input_activation_n_bits" to match "activation_n_bits" if not defined.
+    if 'activation_n_bits' in edit_params_dict and 'supported_input_activation_n_bits' not in edit_params_dict:
```
Comment (ofirgo) [inline tests/common_tests/helpers/generate_test_tp_model.py:34]: use const from mct/constants
and add supported_input_activation_n_bits to consts as well

File: tests/keras_tests/feature_networks_tests/feature_networks/activation_16bit_test.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        x = tf.multiply(inputs, inputs)
+        x = tf.add(x, np.ones((3,), dtype=np.float32))
+        x1 = tf.subtract(x, np.ones((3,), dtype=np.float32))
+        x = tf.multiply(x, x1)
+        x = tf.keras.layers.Conv2D(3, 1)(x)
+        outputs = tf.divide(x, 2*np.ones((3,), dtype=np.float32))
+        return keras.Model(inputs=inputs, outputs=outputs)
+
+    def compare(self, quantized_model, float_model, input_x=None, quantization_info=None):
+        mul1_act_quant = quantized_model.layers[3]
```
Comment (ofirgo) [inline tests/keras_tests/feature_networks_tests/feature_networks/activation_16bit_test.py:50]: use get_layers_from_model_by_type instead of indices

File: tests/keras_tests/feature_networks_tests/feature_networks/activation_16bit_test.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+                                  "1st mul activation should be forced by TPC to be signed, even though activations as all positive.")
+        self.unit_test.assertTrue(mul2_act_quant.activation_holder_quantizer.num_bits == 8,
+                                  "2nd mul activation bits should be 8 bits because of following div node.")
+
+
+class Activation16BitMixedPrecisionTest(Activation16BitTest):
+
+    def get_tpc(self):
+        tpc = mct.get_target_platform_capabilities(TENSORFLOW, IMX500_TP_MODEL, 'v4')
+        mul_op_set = get_op_set('Mul', tpc.tp_model.operator_set)
```
Comment (ofirgo) [inline tests/keras_tests/feature_networks_tests/feature_networks/activation_16bit_test.py:64]: why do you need all this? isn't 16 bit activation should be already in the config options of these layers?

File: tests/keras_tests/feature_networks_tests/test_features_runner.py [hunk 12]
```diff
@@ -788,9 +790,14 @@ def test_keras_tpcs(self):
         TpcTest(f'{C.IMX500_TP_MODEL}.v2_lut', self).run_test()
         TpcTest(f'{C.IMX500_TP_MODEL}.v3', self).run_test()
         TpcTest(f'{C.IMX500_TP_MODEL}.v3_lut', self).run_test()
+        TpcTest(f'{C.IMX500_TP_MODEL}.v4', self).run_test()
         TpcTest(f'{C.TFLITE_TP_MODEL}.v1', self).run_test()
         TpcTest(f'{C.QNNPACK_TP_MODEL}.v1', self).run_test()
 
+    def test_16bit_activations(self):
```
Comment (ofirgo) [inline tests/keras_tests/feature_networks_tests/test_features_runner.py:797]: consider adding tests for different scenarios where 16bit candidates are filtered because of graph incompatibility (if there are other edge cases then test them as well)

Comment (ofirgo) [inline tests/keras_tests/feature_networks_tests/test_features_runner.py:797]: Also I would add function tests checking the set config function separately for filtered candidates (not a part of a complete MST execution)

File: tests/pytorch_tests/model_tests/test_feature_models_runner.py [hunk 13]
```diff
@@ -648,6 +650,10 @@ def test_torch_tpcs(self):
         TpcTest(f'{C.TFLITE_TP_MODEL}.v1', self).run_test()
         TpcTest(f'{C.QNNPACK_TP_MODEL}.v1', self).run_test()
 
+    def test_16bit_activations(self):
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/test_feature_models_runner.py:653]: same comments as in keras

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/target_platform_capabilities/target_platform/op_quantization_config.py [hunk 14]
```diff
@@ -114,11 +114,13 @@ def __init__(self,
                  attr_weights_configs_mapping: Dict[str, AttributeQuantizationConfig],
                  activation_quantization_method: QuantizationMethod,
                  activation_n_bits: int,
+                 supported_input_activation_n_bits: Union[int, Tuple[int]],
```
Comment (ofirgo) [inline model_compression_toolkit/target_platform_capabilities/target_platform/op_quantization_config.py:117]: didn't we say that you give this argument a default value as well? (8)

**Review by ofirgo — CHANGES_REQUESTED** [review]
Just one small question

**Review by ofirgo — APPROVED** [review]
